### PR TITLE
ch4: introduce am_in_use flag in MPIR_Request

### PR DESCRIPTION
**Pull request by srirajpaul**
The am_in_use flag tells whether active message is associated with current
request. MPID_Request_complete() uses this flag to release AM related
allocation.

Co-authored by: Chongxiao Cao <chongxiao.cao@intel.com>

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

File: src/mpid/ch4/src/ch4_request.h [hunk 1]
```diff
@@ -89,7 +89,7 @@ MPL_STATIC_INLINE_PREFIX int MPID_Request_complete(MPIR_Request * req)
         if (req->completion_notification)
             MPIR_cc_dec(req->completion_notification);
 
-        if (MPIDIG_REQUEST(req, req)) {
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_request.h:92]: Why checking whether `MPIDIG_REQUEST(req, req)` is `NULL` isn't sufficient?

Comment (srirajpaul) [reply src/mpid/ch4/src/ch4_request.h:92]: `MPIDIG_REQUEST(req, req)` is `NULL `is a code smell.

Comment (srirajpaul) [reply src/mpid/ch4/src/ch4_request.h:92]: I mean checking whether the value is `NULL ` for other purposes.

**Comment by srirajpaul** [comment]
test:mpich/ch4/ofi

File: src/mpid/ch4/include/mpidpre.h [hunk 2]
```diff
@@ -282,6 +282,10 @@ typedef struct MPIDI_Devreq_t {
     struct MPIR_Request *anysource_partner_request;
 #endif
 
+    /* Used to track AM request is allocated.
+     * True: ch4.am is being used.
+     * False: ch4.am is not being used. */
+    bool am_in_use;
     union {
```
Comment (hzhou) [inline src/mpid/ch4/include/mpidpre.h:289]: There are quite a few members in this union. Use a single `am_in_use` still takes in assumptions. I think the proper code is to have an enum, e.g. `ch4_req_type` to precisely identify which part of the union is in effect.

Comment (srirajpaul) [reply src/mpid/ch4/include/mpidpre.h:289]: Good point. Looks like, `am_in_use `is not tracking AM specifically. It is just tracking whether the `req `is there or not. One way to resolve it is change the name e.g: `req_in_use`.  But if are only trying to find whether `req` is allocated or not, then checking `MPIDI_REQUEST(req,req)` in `NULL ` is also fine. Let me look into it a little more and get back.

**Comment by srirajpaul** [comment]
Seems `am_in_use `is not adding any functionality more than` MPIDI_REQUEST(req,req)` is `NULL `check. So dropping the PR.

**Pull request closed without merging**